Support vmap + custom autograd function/improve DTensor constructor inefficiency

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #162240

This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. 

Since there are number of requests lately to add various operators in pre-dispatch IR, I introduce a decorator in export that works similar to `allow_in_graph`. Basically:
1) We intercept custom_autograd_function.apply at pre-dispatch mode when this decorator is applied 
2) We apply `flat_apply` HOP to hide the pytree spec for this autograd function. Note that this adds restriction that this custom autograd function needs to take in fx-able types.
3) subclass constructor decorator is implemented similarly, so we just refactor it to use similar implementation as this new decorator. eventually we should delete the subclass constructor decorator. 
4) Move some code in subclass constructor decorator to exit early in non-export environment which should shave off some inefficiency (around 1% according to @swolchok 's benchmark) 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Differential Revision: [D82141316](https://our.internmc.facebook.com/intern/diff/D82141316)

tugsbayasgalan committed:
Support vmap + custom autograd function

[ghstack-poisoned]

zou3519 commented on torch/export/custom_ops.py:
```diff
@@ -24,3 +26,23 @@ def _access_subclass_inner_tensor(
+
+def _call_custom_autograd_function_in_pre_dispatch(function_cls_name, *args, **kwargs):
+    """
+    Import a custom autograd function by string name and call it. This is pretty bad
+    because:
```
you're just yolo-ing a custom autograd.Function into pre-dispatch IR?

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. This can only be triggered in export pre-dispatch mode, we just register an implementation inside export. I had to do some janky stuff:
1. I create python-only export function that takes in custom autograd function name and args and internally imports the autograd function. I had to do this because there is no schema for CustomAutogradFunction.apply so we can't use custom ops library. 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. This can only be triggered in export pre-dispatch mode, we just register an implementation inside export. I had to do some janky stuff:
1. I create python-only export function that takes in custom autograd function name and args and internally imports the autograd function. I had to do this because there is no schema for CustomAutogradFunction.apply so we can't use custom ops library. 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. This can only be triggered in export pre-dispatch mode, we just register an implementation inside export. I had to do some janky stuff:
1. I create python-only export function that takes in custom autograd function name and args and internally imports the autograd function. I had to do this because there is no schema for CustomAutogradFunction.apply so we can't use custom ops library. 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/wrappers.py:
```diff
@@ -188,6 +220,9 @@ def wrapper(*args, **kwargs):
             )
         constructor_subclass(*args, **kwargs)
 
+        if not torch.compiler.is_exporting():
+            return
```
Sure.

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. This can only be triggered in export pre-dispatch mode, we just register an implementation inside export. I had to do some janky stuff:
1. I create python-only export function that takes in custom autograd function name and args and internally imports the autograd function. I had to do this because there is no schema for CustomAutogradFunction.apply so we can't use custom ops library. 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D82141316](https://our.internmc.facebook.com/intern/diff/D82141316)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82141316).

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function/improve DTensor constructor inefficiency"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. 

Since there are number of requests lately to add various operators in pre-dispatch IR, I introduce a decorator in export that works similar to `allow_in_graph`. Basically:
1) We intercept custom_autograd_function.apply at pre-dispatch mode when this decorator is applied 
2) We apply `flat_apply` HOP to hide the pytree spec for this autograd function. Note that this adds restriction that this custom autograd function needs to take in fx-able types.
3) subclass constructor decorator is implemented similarly, so we just refactor it to use similar implementation as this new decorator. eventually we should delete the subclass constructor decorator. 
4) Move some code in subclass constructor decorator to exit early in non-export environment which should shave off some inefficiency (around 1% according to swolchok 's benchmark) 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D82141316](https://our.internmc.facebook.com/intern/diff/D82141316)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82141316).

tugsbayasgalan committed:
Update on "Support vmap + custom autograd function/improve DTensor constructor inefficiency"


This makes gemma3 exportable on transformers=4.55.4

In HF, there is a torch funciton mode called TransformGetItemToIndex which internally calls custom autograd function. When this custom autograd function is called under vmap, It triggers CustomFunctionHigherOrderOP which error-ed because there was no pre-dispatch proxy mode implementation. 

Since there are number of requests lately to add various operators in pre-dispatch IR, I introduce a decorator in export that works similar to `allow_in_graph`. Basically:
1) We intercept custom_autograd_function.apply at pre-dispatch mode when this decorator is applied 
2) We apply `flat_apply` HOP to hide the pytree spec for this autograd function. Note that this adds restriction that this custom autograd function needs to take in fx-able types.
3) subclass constructor decorator is implemented similarly, so we just refactor it to use similar implementation as this new decorator. eventually we should delete the subclass constructor decorator. 
4) Move some code in subclass constructor decorator to exit early in non-export environment which should shave off some inefficiency (around 1% according to swolchok 's benchmark) 


Fixes: https://github.com/pytorch/pytorch/issues/161563#issuecomment-3246309758 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D82141316](https://our.internmc.facebook.com/intern/diff/D82141316)

[ghstack-poisoned]
